File: src/Easy/q121.java
package Easy;

public class q121 {
    public static int maxProfit(int[] prices) {
        int lowestPrice = prices[0];
        int maxProfit = 0;

        for (int i=0;i<prices.length;++i) {
            maxProfit = Math.max(maxProfit, prices[i] - lowestPrice);
            lowestPrice = Math.min(lowestPrice, prices[i]);
        }

        return maxProfit;
    }

    public static void testSample() {
        System.out.println(maxProfit(new int[]{7,1,5,3,6,4}));
    }
}

File: Python/src/easy/q121.py
from typing import List

class Solution:
    """
    Leetcode 121: Best Time to Buy and Sell Stock
    """

    def maxProfit(self, prices: List[int]) -> int:
        '''Kanade's algorithm.'''
        buy_price = prices[0]
        profit = 0

        for price in prices:
            if price < buy_price: # if current price < lowest buying price, update
                buy_price = price
            if price - buy_price > profit: # if profit from buying stock at lowest point and sell today > max profit, update
                profit = price - buy_price

        return profit